src/main/kotlin/org/rust/ide/typing/paste/RsConvertJsonToStructCopyPasteProcessor.kt

```diff
@@ -196,22 +196,113 @@ private fun createAndInsertStruct(
+
+// foo_bar_baz
+private fun String.isSnakeCase(): Boolean = isNotEmpty() && none { it.isUpperCase() || it == '-' }
+
+// foo-bar-baz
+private fun String.isKebabCase(): Boolean = isNotEmpty() && none { it.isUpperCase() }
+
+// FOO_BAR_BAZ
+private fun String.isScreamingSnakeCase(): Boolean = isNotEmpty() && none { it.isLowerCase() || it == '-' }
+
+// FOO-BAR-BAZ
+private fun String.isScreamingKebabCase(): Boolean = isNotEmpty() && none { it.isLowerCase() }
+
+private const val DEFAULT_CONVENTION = "snake_case";
+
+/**
+ * Tries to guess the naming convention from the field name.
+ * If the convention matches snake case, no renaming will be applied, since it is the default.
+ */
+private fun getNamingConvention(field: String): String = when {
+    // The order of the checks is important
+    field.isSnakeCase() -> DEFAULT_CONVENTION
+    field.isKebabCase() -> "kebab-case"
+    field.isScreamingSnakeCase() -> "SCREAMING_SNAKE_CASE"
+    field.isScreamingKebabCase() -> "SCREAMING-KEBAB-CASE"
+    field.isPascalCase() -> "PascalCase"
+    field.isCamelCase() -> "camelCase"
+    else -> DEFAULT_CONVENTION
+}
+
+private fun StringBuilder.writeStructField(
+    field: String,
+    type: DataType,
+    nameMap: Map<Struct, String>,
+    mostCommonConvention: String,
+) {
+    val normalizedName = normalizeFieldName(field)
+    val serdeType = getSerdeType(type, nameMap)
+    val convention = getNamingConvention(field)
+    if (convention != mostCommonConvention) {
+        append("#[serde(rename = \"$convention\")]\n")
+    }
+
+    append("pub $normalizedName: ${serdeType},\n")
+}
+
+/**
+ * Finds the most common naming convention if there are any fields.
+ */
+private fun getMostCommonConvention(struct: Struct): String? {
+    return struct.fields.keys
+        .groupingBy { getNamingConvention(it) }
+        .eachCount()
+        .toList()
+        .sortedWith { a, b ->
+            // Sort in descending order by count
+            // If the count is the same, prefer the default convention
+            // Otherwise sort by convention name
+            if (a.second != b.second) {
+                b.second.compareTo(a.second)
+            } else if (a.first == DEFAULT_CONVENTION) {
+                -1
+            } else if (b.first == DEFAULT_CONVENTION) {
+                1
+            } else {
+                a.first.compareTo(b.first)
+            }
+        }
+        .firstOrNull()?.first
+}
+
+private fun generateStruct(
     factory: RsPsiFactory,
     struct: Struct,
     nameMap: Map<Struct, String>,
     hasSerdeDependency: Boolean
 ): RsStructItem? {
     val structString = buildString {
+        val mostCommon = getMostCommonConvention(struct) ?: DEFAULT_CONVENTION
+
         if (hasSerdeDependency) {
             append("#[derive(Serialize, Deserialize)]\n")
         }
+        if (mostCommon != DEFAULT_CONVENTION) {
+            append("#[serde(rename_all = \"$mostCommon\")]\n")
```

Comment:
Let's use `rename_all` only if struct has more then one field (and use single `rename` for the field otherwise)